gsy_framework/sim_results/carbon_emissions.py

```diff
@@ -0,0 +1,246 @@
+            pd.to_numeric, errors="coerce"
+        )
+        df_numeric.index = pd.to_datetime(df_numeric.index)
+        df_numeric["Total Energy (kWh)"] = df_numeric.sum(axis=1) * 1000  # convert MWh to kWh
+        df_numeric["Total Carbon (gCO2eq)"] = df_numeric.mul(emissions_map, axis=1).sum(axis=1)
+        df_numeric["Ratio (gCO2eq/kWh)"] = (
+            df_numeric["Total Carbon (gCO2eq)"] / df_numeric["Total Energy (kWh)"]
+        )
+        return df_numeric[["Ratio (gCO2eq/kWh)"]]
+
+    def calculate_carbon_emissions_from_imported_energy(
+        self, df_carbon_ratio: pd.DataFrame, df_imported_energy: pd.DataFrame
+    ):
+        """Calculate the carbon emissions based on the imported/exported energy"""
+
+        df_carbon_ratio.index = pd.to_datetime(df_carbon_ratio.index)
+        df_imported_energy["simulation_time"] = pd.to_datetime(
+            df_imported_energy["simulation_time"]
+        )
+        df_combined = pd.merge(
+            df_imported_energy,
+            df_carbon_ratio,
+            left_on="simulation_time",
+            right_index=True,
+            how="left",
+        )
+        df_combined["carbon_generated_base_case"] = (
+            df_combined["imported_from_community"] + df_combined["imported_from_grid"]
+        ) * df_combined["Ratio (gCO2eq/kWh)"]
+        df_combined["carbon_generated_gsy"] = (
+            df_combined["imported_from_grid"] * df_combined["Ratio (gCO2eq/kWh)"]
+        )
+        df_combined["carbon_savings"] = (
+            df_combined["carbon_generated_base_case"] - df_combined["carbon_generated_gsy"]
+        )
+        df_combined = df_combined[
+            [
+                "simulation_time",
+                "imported_from_community",
+                "imported_from_grid",
+                "carbon_generated_base_case",
+                "carbon_generated_gsy",
+                "carbon_savings",
+            ]
+        ]
+        return df_combined
+
+    def update(self, country_code: str, time_slot: duration, results_list: List[Dict]):
+        """Update the results_dict with carbon generated and carbon savings"""
+
+        start_date = pd.Timestamp(results_list[0]["current_market"].split("T")[0])
+        end_date = pd.Timestamp(results_list[-1]["current_market"].split("T")[0])
+        df_carbon_ratio = self.calculate_carbon_ratio(country_code, start_date, end_date)
+        df_carbon_ratio.to_csv("carbon_ratio.csv")
+        for result in results_list:
+            current_market = result["current_market"]
+
+            for area_uuid in result["results"].keys():
+                area_result = result["results"][area_uuid]
+                target_timestamp = pd.Timestamp(current_market).tz_localize("UTC")
+
+                nearest_index = df_carbon_ratio.index.get_indexer(  # the closest timestamp to the target_timestamp
+                    [target_timestamp], method="nearest"
+                )[
+                    0
+                ]
```

Comment:
IMO the time resolution of the carbon_ratio should not be calculated here but right after retrieving the data from ENTSOE, essentially the profiles from ENTSOE should be converted to timeseries adhering to the configuration slot_length (therefore, if slot_length=15, 00:00, 00:15, 00:30, 00:45 etc) right at the time that they are first received, and not rely on the "closest timestamp".

Reply:
Hmm... I expect ENTSOE data has the same slot_length in every request. What does not have the same slot_length is our results. So for every result I am finding the closest carbon_ratio data point. But I could indeed already provide the carbon_ratio dataframe with the right slot_length by duplicating the rows or applying an average depending on the slot_length. After I integrate with the gsy-e cli, I will see if using this method still makes sense, and if yes, I will proceed with your suggestion.